### DISCUSSION TITLE
DataCollection access with correct type without phpdocs

### DISCUSSION THREAD

**Discussion by iamrgroot** [spatie/laravel-data#90] [Q&A] [answered]
I'm using phpstan, so I have to use the following phpdoc when I use a DataCollection

```php
class Test extends Data
{
    public function __construct(
        /** @var DataCollection<TestChild> */
        public DataCollection $children,
    ) {
    }
}
```

However,  the ArrayAccess offsetGet function always returns a Data object. So the following:

```php
/** @var Test $test */
$child = $test->children[0];
```
will result in the $child being type Data instead of TestChild, as the offsetGet always returns Data:
```php
    public function offsetGet($offset): Data
    {
        return $this->items->offsetGet($offset);
    }
```

Is there any way this can be automatically resolved to the correct data class? Right now I have to add a phpdoc to tell my IDE it is not a Data object but a TestChild object in order to get autocomplete:

```php
/** @var TestChild $child */
$child = $test->children[0];
```

**Comment by rubenvanassche** [comment, marked as answer]
Hi @IAmRGroot, this actually requires generics within DataCollection, something we're working on for the v2 release!
